Implement new error notifications design

1zaman opened the pull request:
https://openedx.atlassian.net/browse/MA-2412

1zaman commented:
@miankhalid, @BenjiLee, this is ready for an initial review.

BenjiLee commented:
<img src="https://cloud.githubusercontent.com/assets/7101723/19750429/f8c6d050-9bbe-11e6-9e6d-0a3c37d32fc1.png" width="200">

BenjiLee commented:
> The AlertDialogFragment has also been updated to include a new static
constructor that takes string resource IDs instead of the actual
resolved strings, **in order to not have an unneeded dependency** on the
Context in it's notification implementation.

Confusing double negative! but I don't know how else to word it.

miankhalid commented:
Observation: Any snackbar that appears is not dismissable by swiping. Is this something that was left out intentionally ?

1zaman commented:
Yeah, that's something I left out initially since I was just working on the
core change. I'll add a change adding a CoordinatorLayout to the root of
all the layouts later.

On Mon, Oct 31, 2016 at 2:36 PM, Mi@N Kh@LiD <notifications@github.com>
wrote:

> Observation: Any snackbar that appears is not dismissable by swiping. Is
> this something that was left out intentionally ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/866#issuecomment-257249996>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEDCjW73lRQhfhzDiDM8y-tbDiQGGaTEks5q5baZgaJpZM4KZln2>
> .
>

miankhalid commented:
Discussion threads are not loading and snackbar + fullscreen error are appearing simultaneously:

<img src="https://cloud.githubusercontent.com/assets/1710804/19854089/e6a629ec-9f8e-11e6-95d4-85e79f31de71.png" width="250" />

miankhalid commented:
I'm holding off on reviewing this PR until this PR #857 is ready for review (cuz the dependency over that PR is too high), since this PR doesn't seem to be upto mark with the acceptance criteria listed on the ticket.

p.s. It might be a good idea to mark this PR as WIP.

1zaman commented:
The persistent Snackbar errors on multiple screens were caused by the `NavigationFragment`, which is added by `BaseFragmentActivity` if the drawer layout is present, but blocked from opening by most implementations. I have removed error notifications from `NavigationFragment` in commit 6d359c090188ecbdf6e9ae5b320be09a470ab184, since there's no proper place for it to be destroyed (as Snackbars should be shown below the drawer).

I have also removed temporary debugging code that I had accidentally checked in, which was causing all the calls to be reported as failing in commit f46f510c2372adc277d1bd45de8f2588a7e4d93f.

BenjiLee commented:
@1zaman I'm not seeing any snackbars.

BenjiLee commented:
@miankhalid @1zaman Should we prioritize the review of https://github.com/edx/edx-app-android/pull/857 over this?

miankhalid commented:
> Should we prioritize the review of #857 over this?

+1 from me.

1zaman commented:
Yeah, I guess it would be better to get #857 merged before this, since then it will be easier to implement this in the Assessments module easily without major changes later.

miankhalid commented:
I've rebased this PR successfully, and tests are passing too.

The issues mentioned above have been tested and I haven't been able to reproduce any. @BenjiLee can you give it a look too ?

I have found a minor issue in Discussion, where persistent connection error appears in the middle of screen. I'll fix that issue in a followup commit, but before that @verdanmahmood suggested to get the branch tested thoroughly by @MirzaMubasharBaig (which he'll be doing tomorrow hopefully).

So, I'll fix all the issues (if anything else surfaces) in one go.

BenjiLee commented:
I'm still seeing the old UI everywhere.

<img src="https://cloud.githubusercontent.com/assets/7101723/22659198/c1ad5984-ec6a-11e6-980d-c98bdbd30287.png" width="200">

miankhalid commented:
@farhan this PR is ready for a first pass.

p.s ignore this Commit # 2b033d6 (Offline support for Handouts and Announcements) infavor of this Commit # 1ce08f1 (Offline mode changes).

farhan commented:
**Snack bar also appearing with full screen blocker**

Steps:
1) Fresh install the app and keep the internet available.
2) Do login in the app.
3) Choose some course from "My Courses" screen and go to its dashboard.
4) On "Course Dashboard" screen, disable the internet.
5) Now navigate to its Outline screen by pressing "Course" tab.
6) You will see full screen blocker as well as snack bar. Snack bar is not required here.

![1](https://cloud.githubusercontent.com/assets/25842457/25802702/ee81a188-340c-11e7-80d1-1dfb52a5d9eb.png)


**Offline full screen blocker not getting hide after successful reloading of page.**

7) Now enable the internet and press 'Reload' button. Full blocker message keep appearing even the page gets reload successfully.

![2](https://cloud.githubusercontent.com/assets/25842457/25802806/6ceb90c4-340d-11e7-8494-ca5884d837e0.png)

farhan commented:
As mentioned on this wiki: https://openedx.atlassian.net/wiki/display/MA/Offline+Errors

"In the login activity, when the app detects that we are offline, an alert dialog is displayed that tells the user there is no connection. This can be dismissed either by hitting OK or getting the back button. **It also goes away if we have a network connection again.**"

Dialog doesn't go away automatically if have a network connection back again.

farhan commented:
**Here are few issues which needs to be addressed.**

1.  Find Courses screen is according to new design like iOS. Its in both cases if user come on it after singing in or without signing in.
2. On My Courses screen, if courses are not cached, iOS is showing full screen blocker message rather than offline snack bar.
3. Load all the courses on My Courses screen, disable internet, dismiss the offline snack bar, navigate to some course dashboard and come back to My Courses screen,  you will not see the snack bar again (it happens on other screens as well) unlike iOS.
4. Disable internet and tap on some video to watch it. On Course unit screen you will see a snackbar which is unrequired and not appearing in iOS as well. Suggestion: This screen needs to be improved in case of non-downloaded videos, full screen blocker is a good option, otherwise it will be a bad user experience.

farhan commented:
**ErrorNotification, SnackbarErrorNotification, DialogErrorNotification & OverlayErrorNotification** classes code indentation and java docs suppose to be improved at best as whole new design implementation revolves around it.
Indentation could be improved by just selection all code and hit formatting short cut key.

farhan commented:
`panel_offfline_access.xml` layout and its resources should be removed as they are not getting used anywhere.

farhan commented:
Getting offline error in top panel when try the change picture in offline mode.
<img src="https://cloud.githubusercontent.com/assets/25842457/26055974/b3d981ec-398c-11e7-960a-74c54bd00d1e.png" width="200">

farhan commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseBaseActivity.java:
```diff
@@ -65,6 +74,8 @@ protected void onCreate(Bundle arg0) {
             bar.setDisplayHomeAsUpEnabled(true);
             bar.setIcon(android.R.color.transparent);
         }
+        errorNotification = new OverlayErrorNotification(progressWheel);
```
In all the places where we have used name `errorNotification` for a field which is initializing with `OverlayErrorNotification`, `overlayErrorNotification`  name will be more clean and easy to understand the purpose.

farhan commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/MyCoursesListFragment.java:
```diff
@@ -126,12 +134,28 @@ public void onLoadFinished(Loader<AsyncTaskResult<List<EnrolledCoursesResponse>>
+                        break;
+                    }
+                    default: {
+                        errorNotification.showError(R.string.network_connected_error,
```
Are not we unconditionally expecting exception cause due to internet connectivity here. Suppose to take error message from `ErrorUtils.getErrorMessage` function.

miankhalid commented:
@farhan All the issues that you reported have been fixed.

Following are the answers to some of the queries listed:

> Dialog doesn't go away automatically if have a network connection back again.

Blocking errors i.e. Dialogs shouldn't go away automatically.

> Find Courses screen is according to new design like iOS. Its in both cases if user come on it after singing in or without signing in.

Find Courses is out of scope for this PR.

> Indentation could be improved by just selection all code and hit formatting short cut key.

Lets do this at the end when you've approved the PR.

> panel_offline_access.xml layout and its resources

I can remove the functions you mentioned, but this layout is still being used in the VideoListActivity.java

miankhalid commented on OpenEdXMobile/res/layout/content_error.xml:
```diff
@@ -4,23 +4,30 @@
-        android:layout_height="match_parent"
-        android:gravity="center"
-        android:orientation="vertical">
+        android:id="@+id/content_error"
```
They are defined in its style i.e. `@style/content_unavailable_layout`.

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/http/notifications/OverlayErrorNotification.java:
```diff
@@ -0,0 +1,163 @@
+import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
+
+/**
+ * An overlay notification error message on top of the content area.
```
`which could also possess error icon and action button.` is an implementation detail and shouldn't be included in the description.

But yeah the description could be improved like:
`An full screen error notification on top of the content area.`

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/http/notifications/OverlayErrorNotification.java:
```diff
@@ -0,0 +1,163 @@
+/**
+ * An overlay notification error message on top of the content area.
+ */
+public class OverlayErrorNotification extends ErrorNotification {
```
Agreed. Lets do this change at the end when the PR gets approved.

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/http/notifications/SnackbarErrorNotification.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    public void hideError() {
+        if (snackbar != null) {
+            snackbar.dismiss();
```
No need of this change after this commit 0fd2e2f8c2f61588affa45c0193801ae87f13045.

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/http/notifications/DialogErrorNotification.java:
```diff
@@ -0,0 +1,48 @@
+                          @NonNull final Icon icon,
+                          @StringRes final int actionTextResId,
+                          @Nullable final View.OnClickListener actionListener) {
+        AlertDialogFragment.newInstance(0, errorResId, null).show(fragmentManager, null);
```
Great catch! We also need a null check in there for `actionListener`.

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/profiles/UserProfileFragment.java:
```diff
@@ -194,6 +201,18 @@ public void showError(@NonNull Throwable error) {
                 viewHolder.profileBodyContent.setVisibility(View.GONE);
                 viewHolder.contentError.getRoot().setVisibility(View.VISIBLE);
                 viewHolder.contentError.contentErrorText.setText(ErrorUtils.getErrorMessage(error, getContext()));
+                viewHolder.contentError.contentErrorAction.setText(R.string.lbl_reload);
```
`OverlayErrorNotification` can't handle this specific case, cuz we need to hide 2 different layouts incase of error (It takes only layout to hide in its constructor).

> If we getR.string.action_not_completed or R.string.app_version_unsupported error message from above function than still we should show action button with 'Reload' label. Thoughts ?

Reload seems fine for, as we are giving the user a way to reload the screen irrelevant of the error.

> And we are always showing wifi icon in content_error_icon view:
app:iconName="fa-wifi"

Let me improve this logic.

miankhalid commented:
Ready for another pass @farhan.

verdan commented:
@farhan can you please make sure to give it another pass?

farhan commented:
"My Courses" screen still shows snackbar rather than full screen message in case when no online or cache data is available to show. Looks like some exceptions in background.

<img src="https://cloud.githubusercontent.com/assets/25842457/26486132/4ca1880a-4213-11e7-9595-b8913b6d0f94.png" width="200">

farhan commented:
App got crashed with following logs

05-26 15:11:40.842 9505-9505/org.edx.mobile E/AndroidRuntime: FATAL EXCEPTION: main
                                                              Process: org.edx.mobile, PID: 9505
                                                              java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.res.AssetManager android.content.Context.getAssets()' on a null object reference
                                                                  at org.edx.mobile.util.FileUtil.loadTextFileFromAssets(FileUtil.java:44)
                                                                  at org.edx.mobile.util.WebViewUtil.getIntialWebviewBuffer(WebViewUtil.java:23)
                                                                  at org.edx.mobile.view.CourseCombinedInfoFragment.populateAnnouncements(CourseCombinedInfoFragment.java:233)
                                                                  at org.edx.mobile.view.CourseCombinedInfoFragment.access$100(CourseCombinedInfoFragment.java:46)
                                                                  at org.edx.mobile.view.CourseCombinedInfoFragment$3.onResponse(CourseCombinedInfoFragment.java:213)
                                                                  at org.edx.mobile.view.CourseCombinedInfoFragment$3.onResponse(CourseCombinedInfoFragment.java:208)
                                                                  at org.edx.mobile.http.callback.ErrorHandlingOkCallback$1.run(ErrorHandlingOkCallback.java:285)
                                                                  at android.os.Handler.handleCallback(Handler.java:739)
                                                                  at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                  at android.os.Looper.loop(Looper.java:158)
                                                                  at android.app.ActivityThread.main(ActivityThread.java:7224)
                                                                  at java.lang.reflect.Method.invoke(Native Method)
                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
05-26 15:11:42.202 9505-9960/org.edx.mobile D/OkHttp: <-- 200 OK https://courses.edx.org/api/user/v1/accounts/Mkhalid (1624ms)


05-26 15:11:42.202 9505-9960/org.edx.mobile D/OkHttp: Allow: GET, PATCH, HEAD, OPTIONS
05-26 15:11:42.202 9505-9960/org.edx.mobile D/OkHttp: Content-Language: en
05-26 15:11:42.202 9505-9960/org.edx.mobile D/OkHttp: Content-Type: application/j

farhan commented:
> I can remove the functions you mentioned, but this layout is still being used in the VideoListActivity.java

I didn't find the usage of 'panel_offline_access.xml' in project.

farhan commented:
@miankhalid i am done with second pass, need few things to address.

miankhalid commented:
Ready for another pass @farhan

farhan commented:
LGTM :+1:

Just last thing, verify the following case with requirement, do we need to fix it ?

1. Login to the app and go to 'My Courses' screen.
2. Disable internet access.
3. Choose some course and go to 'Course Outline' screen you will see the full screen error message here. Enable the internet and reload the contents.
4. Now go back to 'My Courses' screen, you will still see Offline Snackbar.

miankhalid commented:
@farhan Fixed the last case you mentioned in 81bc08d. Final pass maybe?

farhan commented:
@miankhalid  Snack bar is not hiding for 'Discussion topics' screen for rest required screens its working fine.

**Steps:**
1. Go to 'Discussion topics' screen of any course and disable internet, you will see offline snackbar on screen.
2. Now navigate to next screen by choosing any topic (or go to home screen)
3. Enable the internet
4. Come back to 'Discussion topics' screen, you will still see the offline snack bar.
